### ISSUE TITLE
[BUG] isinstance(obj, TypeA | TypeB) returns wrong result

### ISSUE THREAD

**Issue by behrenhoff** [cython/cython#6420]
### Describe the bug

Cythonized code returns wrong results on isinstance calls with union types.

This following code shows the problem:

```
def cython_check(obj: object) -> bool:
    # this returns the wrong result
    return isinstance(obj, str | bytes)

def cython_check2(obj: object) -> bool:
    # this works correctly
    return isinstance(obj, (str, bytes))
```

When compiled, I get:
```
>>> from my_cythonized_package import cython_check, cython_check2
>>> isinstance("hallo", str | bytes)
True
>>> cython_check("hallo")
False
>>> cython_check2("hallo")
True
```

This issue is certainly related to https://github.com/cython/cython/issues/4631

The issue is that cython's behavior deviates from pure Python's behavior.
I would be happy if cython told me "not yet supported" and failed with a compile error. But returning the wrong result with no error message is not really optimal...

### OS

Linux

### Python version

3.12.3

### Cython version

3.0.11

**Comment by da-woods** [comment]
I don't think we look at abstract base classes at all in the `isinstance` lookup (which I think this is just an audiobook a special case of). Which is a shame because it'll make the check more complicated.

**Comment by scoder** [comment]
I'll look into optimising these type checks. Should be easy enough.

**Comment by da-woods** [comment]
> I'll look into optimising these type checks. Should be easy enough.

Yeah - I think you just need to ensure that both the operands are exact `type`. (So basically just the same rules as for `tuple`).

I think that this fix (or something close) is still the right thing to do even if you can subsequently optimize it more.

**Comment by da-woods** [comment]
By "this fix" I mean https://github.com/cython/cython/pull/6932... didn't spot I was commenting on the issue not the PR

**Comment by scoder** [comment]
3.1.x commit is https://github.com/cython/cython/commit/45dc72dcbe2ea4984dd68d4e4dfa8a992f2121b3
